src/EventGridExtension/OutputBinding/EventGridAsyncCollector.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.EventGrid;
+using Microsoft.Azure.EventGrid.Models;
+
+namespace Microsoft.Azure.WebJobs.Extensions.EventGrid
+{
+    public sealed class EventGridAsyncCollector : IAsyncCollector<EventGridEvent>
+    {
+        // use IEventGridClient for mocking test
+        private readonly IEventGridClient _client;
+        private readonly IList<EventGridEvent> _eventsToSend = new List<EventGridEvent>();
+        private readonly string _topicHostname;
+
+        private ManualResetEventSlim _canAdd = new ManualResetEventSlim(true);
```

Comment:
I would say we drop synchronization for this commit, since it needs to be done for all output bindings,
we don't want each one to have their own implementation, instead, we can introduce a new parameter types called `ConcurrentAsyncCollector`, which is a wrapper over `IAsyncCollector`, and user can decide if they want the thread safety and the lock overhead:
![image](https://user-images.githubusercontent.com/6531400/42254052-9f1e83f2-7ef9-11e8-8dbb-b682bfa1fd9a.png)
looking at the example: [`SyncAsyncCollectorAdapter`](https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Host/Bindings/AsyncCollector/SyncAsyncCollectorAdapter.cs) we can probably have locks over flush and add method, what you think @brandonh-msft

Reply:
Had a chat with @watashiSHUN offline. We still want to move forward with the synchronization here to make this safe. Adding a couple of comments to the approach.